src/equal.c

```diff
@@ -137,14 +137,18 @@ static int cpl_equal_scalar(const Rcomplex* x, const Rcomplex* y, bool na_equal)
     return real_equal && imag_equal;
   }
 }
+
+static int chr_equal_scalar_impl(const SEXP x, const SEXP y) {
+  return !strcmp(Rf_translateCharUTF8(x), Rf_translateCharUTF8(y));
```

Comment:
Maybe do a fast pointer comparison path if `getCharCE()` is the same for both inputs?